DIPs/DIP1xxx-YS.md

```diff
@@ -0,0 +1,142 @@
+| Status:         | Draft                                                                |
+
+## Abstract
+
+Named arguments adds a way to annotate the arguments passed to functions. It makes the intention of passed arguments more clear, which should results in code been more readable.
+
+### Reference
+
+Various solutions have been suggested in the NewsGroup,  [1](https://forum.dlang.org/post/khcalesvxwdaqnzaqotb@forum.dlang.org) and [2](https://forum.dlang.org/post/n8024o$dlj$1@digitalmars.com).
+
+Library only solutions have also been attempted multiple times, see [1](https://forum.dlang.org/post/awjuoemsnmxbfgzhgkgx@forum.dlang.org) and [2](https://github.com/CyberShadow/ae/blob/master/utils/meta/args.d).
+
+There are also a couple DIPs proposed or in development, see [DIP88](https://wiki.dlang.org/DIP88) and [rikkimax](https://github.com/rikkimax/DIPs/blob/named_args/DIPs/DIP1xxx-RC.md)
+
+## Contents
+* [Rationale](#rationale)
+* [Description](#description)
+* [Breaking Changes and Deprecations](#breaking-changes-and-deprecations)
+* [Reviews](#reviews)
+* [Appendix](#appendix)
+
+## Rationale
+
+Named arguments is a common programming language feature found in various dynamic or compiled languages. There are various motivation behind that, such as to make code more readable, and to pass arguments in arbitrary order.
+
+Unlike named arguments in other languages and previous DIPs, this DIP only aim to address the readability issue. This is to increase the chance of adaptation, while still keeping one of the major niceties of this feature.
+
+Consider this example:
+
+```d
+// What are these arguments?
+DecimalNumber product = CalculateProduct(values, 7, false, null);
+```
+
+It is really difficult to decipher what each argument actually means. This is a real existing problem. As even Google's C++ coding style guide tries to provide a solution for that (they suggest using comments, we will discuss later).
+
+This DIP also has the added benefits of preventing silent breakage when the meanings of function parameters changed, but their type don't (assuming the meanings are reflected in the names). For example:
+
+```d
+// Previous API: void draw_rect(int x, int y, int width, int height);
+// New API:
+void draw_rect(int x0, int y0, int x1, int y1);
+```
+
+## Description
+
+The bird's eye view of this change is simple. In function calls, the arguments passed can be prefixed with a name, like this: `draw_rect(x: 0, y: 0, width: 1, height: 1)`. And when names mismatch, errors will be generated.
+
+However, there are quite a bit of details that need to be nailed down.
+
+### Ordering of arguments
+
+As stated before, this DIP only tries to address the readability issue, therefore passing arguments in different order than in the function definition is not allowed. All arguments have to be passed in the same order as defined in function parameter list, optionally they can be prefixed with a name.
```

Comment:
This will likely make the feature unpopular and underused. If you want to avoid technical corner cases, a better option will likely be that reordering can only happen if all parameters are named.

Reply:
Reordering is quite useful if there are many arguments with default values, I think C# does this for the Win API. You don't want to specify the value for all those default values just because you want to override the last value.

Reply:
This proposal is intended to be minimal but extensible. Although this proposal itself doesn't include argument reordering, it does not prohibit that from being added in a future DIP.

Reply:
Reordering is going to interfere with overload resolution, which will blow up the scope of this DIP.

For example

```d
// Imagine int and double are two incompatible types
int fun1(int a, double b);
int fun1(double b, int a);

fun1(a: 1, b: 1.0);
```

Without reordering, it's clear what fun1 calls

Reply:
You can address that by simply stating that your example will trigger a compilation error due to ambiquity. Since named parameters are currently not allowed, it does not mean any breakage.

If your DIP as it's now get's implemented, we can't change it that way anymore without breakage.

Reply:
> we can't change it that way anymore without breakage.

Good point. I will think about this. Looks like in the DIP we will only be able to support reordering if all arguments are labeled.

Reply:
C# also errors if call is ambiguous: http://rextester.com/VQAQ47472.